src/Sentry/MetricAggregator.cs

```diff
@@ -0,0 +1,273 @@
+                if (!shutdownRequested)
+                {
+                    try
+                    {
+                        var delay = Task.Delay(_flushInterval, shutdownTimeout.Token).ConfigureAwait(false);
+                        await Task.Delay(_flushInterval, shutdownTimeout.Token).ConfigureAwait(false);
+                    }
+                    // Cancellation requested and no timeout allowed, so exit even if there are more items
+                    catch (OperationCanceledException) when (_options.ShutdownTimeout == TimeSpan.Zero)
+                    {
+                        _options.LogDebug("Exiting immediately due to 0 shutdown timeout.");
+
+                        shutdownTimeout.Cancel();
+
+                        return;
+                    }
+                    // Cancellation requested, scheduled shutdown
+                    catch (OperationCanceledException)
+                    {
+                        _options.LogDebug(
+                            "Shutdown scheduled. Stopping by: {0}.",
+                            _options.ShutdownTimeout);
+
+                        shutdownTimeout.CancelAfterSafe(_options.ShutdownTimeout);
+
+                        shutdownRequested = true;
+                    }
+                }
+
+                // Work with the envelope while it's in the queue
+                foreach (var key in GetFlushableBuckets())
+                {
+                    // TODO: Check if a shutdown request has been made
+                    try
+                    {
+                        _options.LogDebug("Flushing metrics for bucket {0}", key);
+                        var bucket = Buckets[key];
+                        _captureMetrics(bucket.Values);
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        _options.LogInfo("Shutdown token triggered. Time to exit.");
+                        return;
+                    }
+                    catch (Exception exception)
+                    {
+                        _options.LogError(exception, "Error while processing metric aggregates.");
+                    }
+                    finally
+                    {
+                        _options.LogDebug("Metric flushed for bucket {0}", key);
+                        Buckets.TryRemove(key, out _);
+                    }
+                }
+            }
+        }
+        catch (Exception e)
+        {
+            _options.LogFatal(e, "Exception in the Metric Aggregator.");
+            throw;
+        }
+    }
+
+    /// <summary>
+    /// Returns the keys for any buckets that are ready to be flushed (i.e. are for periods before the cutoff)
+    /// </summary>
+    /// <returns>
+    /// An enumerable containing the keys for any buckets that are ready to be flushed
+    /// </returns>
+    internal IEnumerable<long> GetFlushableBuckets()
+    {
+        var cutoff = MetricBucketHelper.GetCutoff();
```

Comment:
Reasoning below

```suggestion
        if (_buckets is null)
        {
            return Enumerable.Empty<long>();
        }
        var cutoff = MetricBucketHelper.GetCutoff();
```

Reply:
Ah, I see... so if metrics are enabled but no metrics have yet been emitted. Cool, good idea (I'll fix up the syntax so that it works with the `yield return` and the `Lazy<T>`.